modules/objdetect/test/test_qrcode.cpp

```diff
@@ -1,74 +1,98 @@
+    String dataset_config = cvtest::TS::ptr()->get_data_path() + "qrcode/dataset_config.json";
+    FileStorage file_config(dataset_config, FileStorage::WRITE);
+
+    file_config << "test_images" << "[";
+    size_t images_count = sizeof(qrcode_images_name) / sizeof(String);
+    for (size_t i = 0; i < images_count; i++)
+    {
+        file_config << "{:" << "image_name" << qrcode_images_name[i];
+        String image_path = root + qrcode_images_name[i];
+        std::vector<Point> transform;
+        Mat src = imread(image_path, IMREAD_GRAYSCALE);
+        EXPECT_TRUE(detectQRCode(src, transform));
+        file_config << "x" << "[:";
+        for (size_t j = 0; j < transform.size(); j++) { file_config << transform[j].x; }
+        file_config << "]";
+        file_config << "y" << "[:";
+        for (size_t j = 0; j < transform.size(); j++) { file_config << transform[j].y; }
+        file_config << "]" << "}";
+    }
+    file_config << "]";
+    file_config.release();
+}
+
+#else
+
+typedef testing::TestWithParam< String > Objdetect_QRCode;
+TEST_P(Objdetect_QRCode, regression)
 {
     String root = cvtest::TS::ptr()->get_data_path() + "qrcode/";
-    // String cascades[] =
-    // {
-        // root + "haarcascade_frontalface_alt.xml",
-        // root + "lbpcascade_frontalface.xml",
-        // String()
-    // };
-
-    // vector<Rect> objects;
-    // RNG rng((uint64)-1);
-
-    // for( int i = 0; !cascades[i].empty(); i++ )
-    // {
-        // printf("%d. %s\n", i, cascades[i].c_str());
-        // CascadeClassifier cascade(cascades[i]);
-        // for( int j = 0; j < 100; j++ )
-        // {
-            // int width = rng.uniform(1, 100);
-            // int height = rng.uniform(1, 100);
-            // Mat img(height, width, CV_8U);
-            // randu(img, 0, 256);
-            // cascade.detectMultiScale(img, objects);
-        // }
-    // }
+    String dataset_config = cvtest::TS::ptr()->get_data_path() + "qrcode/dataset_config.json";
+    FileStorage file_config(dataset_config, FileStorage::READ);
+
+    if (file_config.isOpened())
+    {
+        String image_path = root + String(GetParam());
+        std::vector<Point> transform;
+        Mat src = imread(image_path, IMREAD_GRAYSCALE);
+        EXPECT_TRUE(detectQRCode(src, transform));
+
+        FileNode images_list = file_config["test_images"];
+        int index = 0, images_count = static_cast<int>(images_list.size());
+        ASSERT_GT(images_count, 0);
+
+        for (; index < images_count; index++)
+        {
+            String test_image = images_list[index]["image_name"];
+            if (test_image == String(GetParam())) { break; }
+        }
+
+        for (int i = 0; i < 4; i++)
+        {
+            float x = images_list[index]["x"][i];
+            float y = images_list[index]["y"][i];
+            EXPECT_LT(abs(x - transform[i].x), 3);
+            EXPECT_LT(abs(y - transform[i].y), 3);
```

Comment:
For floats use `fabs`.

It is better to use `EXPECT_NEAR(v1, v2, abs_error)` for that.

Reply:
done, changed EXPECT_NEAR